### DISCUSSION TITLE
T2i Semantic Segmentation Color Reference Chart - v21

### DISCUSSION THREAD

**Discussion by AugmentedRealityCat** [Mikubill/sd-webui-controlnet#503] [Show and tell]
EDIT: a [new version with proper HEX codes has been posted over here](https://github.com/Mikubill/sd-webui-controlnet/discussions/503#discussioncomment-13890552).

This document presents the colors associated with the 182 classes of objects recognized by the T2i Semantic Segmentation model.

This version (v21) is complete and all data has been cross-checked against the official code, thanks to [u/Doggettx](https://www.reddit.com/user/Doggettx/) .

Any previous version should be deleted as they were containing errors and there was also missing data for some classes.

pdf version
[T2i_Segmentation_Color_Ref_v21.pdf](https://github.com/TencentARC/T2I-Adapter/files/10892954/T2i_Segmentation_Color_Ref_v21.pdf)

spreadsheet in LibreOffice ods format
[T2i_Segmentation_Color_Ref_v21.ods](https://github.com/TencentARC/T2I-Adapter/files/10892970/T2i_Segmentation_Color_Ref_v21.ods)

![T2i_Segmentation_Color_Ref_v21](https://user-images.githubusercontent.com/114955776/222986602-8416e4c9-c834-4657-86eb-ed10c69523a2.png)

**Comment by if-ai** [comment]
Does exist a Sheet for the openpose bone colors like this one. I tried color picking the oficial documentation for my free rig here https://ko-fi.com/impactframes but then the model was coming up backwards. I notice the Color in the posex extension were different

**Reply by AugmentedRealityCat** [reply]
I haven's seen such a chart yet. If you think it would be useful, I can build one. It certainly won't take as much time as the T2i one !

**Reply by if-ai** [reply]
Thank you. Yes, It would be helpful for the people that is building OpenPose for SD if you could make one. At the moment the specification are not clear and I know for sure that controlnet is not using the official color scheme at least for the head. 

It would be great if we had an official color scheme for the extension so we can all know we are using the correct colors specially if there are plans of training a new model that include hands and feet the future.

**Reply by AugmentedRealityCat** [reply]
Here is my Work-In-Progress. You are the first to see it ! Please test and let me know if you see any errors, mistakes or missing information.
If all is good I'll make a proper post with it.

EDIT: REMOVED old inaccurate version, please refer to the new version posted below.

**Reply by if-ai** [reply]
Thank you, you are super fast, I will check the colors, They look accurate and better than my PS color picker but the head is different on the controlNet model because the times I tried using blue for the left ear, it yields most of the poses with the back at the viewer. And if you see all of the openPose bones extensions they have red color for those joints.

**Reply by if-ai** [reply]
I have asked here maybe we get lucky and someone knows the answer https://github.com/lllyasviel/ControlNet/discussions/266

**Reply by AugmentedRealityCat** [reply]
Here is version 13. 

[Openpose-18-keypoints_coco_color_codes_v13.pdf](https://github.com/Mikubill/sd-webui-controlnet/files/10964844/Openpose-18-keypoints_coco_color_codes_v13.pdf)

![Openpose-18-keypoints_coco_color_codes_v13](https://user-images.githubusercontent.com/114955776/224917086-e66b42be-83b0-4d4a-8f63-0dc21889e9fa.png)

Color data has been updated from first published version (v10), and the whole thing as been reordered to better reflect how everything is numbered in the code. Another fix is that bone links (line) colors are now blended at 60% over black, exactly like the preprocessor does, while the previous version was blended at 70%, (which is the recipe used by the PoseX extension).

This means that if you want to color bones (the lines) with alpha transparency instead of opaque colors, then you can use the corresponding joint color (the bright ones on the left side of the chart) and blend it at 60%. Over black, that will give you exactly the result shown on the right side of the chart. 

Keep in mind that joint colors (the little circles) must be kept completely opaque.

**Reply by if-ai** [reply]
It seems to work well, at a minimum works better than before. 
The right hand looks like it has a little issue as it is not guiding as it should, I might need to run more tests to be sure, but so far is great.
also the neck bone is missing from the chart, but it is not giving me trouble at all.

![bone0002](https://user-images.githubusercontent.com/21185218/225479778-f549927f-42a6-40c2-b206-41bd3c41ff80.png)
![04989-2503043067-Beautifully illustrated scene of Spiderman enquires the hotdog stand attendant about a nearby suspect, taking in the breathtakin](https://user-images.githubusercontent.com/21185218/225479789-ad294734-5353-4d29-a082-ee2e26cebea7.png)
![bone00002](https://user-images.githubusercontent.com/21185218/225479848-26a2078e-f257-443c-81f0-db862b2154e4.png)
![04980-2247828476-Beautifully illustrated scene of Spiderman swings passes into midair in front of the iconic skyscraper buildboard with his webs,](https://user-images.githubusercontent.com/21185218/225479888-2f54b1f6-479b-4f2e-821e-65f616f1ad4b.png)

![bone0000](https://user-images.githubusercontent.com/21185218/225479898-08da0878-a49c-421e-a264-3f9b821bd19f.png)
![04978-2247828476-Beautifully illustrated scene of Spiderman jumping midair into a ball out of window of iconic skyscraper window with his webs, t](https://user-images.githubusercontent.com/21185218/225479922-00fbc926-894b-4e92-9457-e738953b7dc3.png)
![04979-2247828476-Beautifully illustrated scene of Spiderman jumping midair into a ball out of window of iconic skyscraper window with his webs, t](https://user-images.githubusercontent.com/21185218/225479931-642e0c62-7c37-41c3-94a2-cabdb968d3d7.png)

**Reply by AugmentedRealityCat** [reply]
>  The right hand looks like it has a little issue as it is not guiding as it should, I might need to run more tests to be sure, but so far is great. 

I would have to find more information about the hand model official colors to understand. I did not include hands at all in my study when I made the chart. Please run more tests and let me know if you find something wrong with the right wrist joint or the forearm bone link colors per se. Maybe by removing the hands and see if there is still a problem without them? I understand you want to use hands, but this is just to test the chart's accuracy before officially releasing it. 

> also the neck bone is missing from the chart, but it is not giving me trouble at all.
It's called `head` in bones list - `neck` is the name used for the joint (point).

Very nice examples by the way. I made many spiderman pictures for a project I was pitching to a client last fall and it was really hard to get anatomically-correct pictures of that character. ControlNet and T2i are really changing that !

Thanks for your feedback.

**Reply by if-ai** [reply]
Understood I will use the heads then, and yes, I meant the forearm but is not a problem I think that is why it looks more orange color in the openpose extension.  Sorry I was so busy I could not answer you before but I needed to retexture the blender model and try it.

Again Thank you for working on this I see the characters guidance has already improve it.

**Reply by hafriedlander** [reply]
Openpose bones don't really have different colors to the points - instead both the bones and the points should be semi-opaque (at around 50-60%). The bones should also be ovals, not just lines. 

(From reference image https://cmu-perceptual-computing-lab.github.io/openpose/web/html/.github/media/keypoints_pose_18.png from  https://cmu-perceptual-computing-lab.github.io/openpose/web/html/doc/md_doc_02_output.html)

I've confirmed that the ControlNet annotator uses the same colors / opacity / bone shape as the above reference by looking at the code, and that's the way I've implemented in gyre.ai too.

(ControlNet reference - https://github.com/lllyasviel/ControlNet/blob/main/annotator/openpose/util.py#L36)

However both the ControlNet and T2I openpose adapter don't seem particularly sensitive to errors in the skeleton - the A1111 extension gets (got? haven't checked in a while) the right hand color completely wrong, and it mostly works fine.

**Reply by AugmentedRealityCat** [reply]
Thanks a lot for confirming that the color chart is accurate ! Your help is greatly appreciated. 

About the transparency, I had discovered that, and the discrepancy between the transparency values from application to application. For this version, I used the transparency shown by the pre-processor, which is 60% , as indicated by the "60% color" label I've put at the top of the joint colors column. I also kept joints and bones that share the same basic color on the same row to help with that association. 

Since this whole transparency and color sharing between joints and bones was not apparent to you when you first looked at the chart though, I have to conclude that I could do a better job explaining that. If you have any suggestion for a better version, I would love to hear about it.

To conclude about the transparency values, I've seen 70% used in an openpose extension for A1111 (I did use that for a previous version of the chart), but I also saw one extension where all the colors were opaque, and it was still working.

As for the hands, there are some examples of what you are saying in the spiderman pictures shown above in this thread, and it was my intention to create a chart for those colors as well. 

Can you give me some clue about where and how to find more data about the openpose hand model colors ? And what the differences should be between the right and the left hand ? 
I've seen this ref picture already but maybe there is a list with RGB values in a repo somewhere : 
![image](https://user-images.githubusercontent.com/114955776/228992314-77772f05-0f4f-4f4a-8664-ebfbaf460e63.png)

Thanks again for your help.

**Reply by if-ai** [reply]
> Openpose bones don't really have different colors to the points - instead both the bones and the points should be semi-opaque (at around 50-60%). The bones should also be ovals, not just lines.
> 
> (From reference image https://cmu-perceptual-computing-lab.github.io/openpose/web/html/.github/media/keypoints_pose_18.png from https://cmu-perceptual-computing-lab.github.io/openpose/web/html/doc/md_doc_02_output.html)
> 
> I've confirmed that the ControlNet annotator uses the same colors / opacity / bone shape as the above reference by looking at the code, and that's the way I've implemented in gyre.ai too.
> 
> (ControlNet reference - https://github.com/lllyasviel/ControlNet/blob/main/annotator/openpose/util.py#L36)
> 
> However both the ControlNet and T2I openpose adapter don't seem particularly sensitive to errors in the skeleton - the A1111 extension gets (got? haven't checked in a while) the right hand color completely wrong, and it mostly works fine.

Thank you so much for clarifying all this. I am going to improve the rig to account for the transparency and oval shapes of the limbs.

**Comment by hafriedlander** [comment]
https://github.com/CMU-Perceptual-Computing-Lab/openpose/blob/master/include/openpose/hand/handParameters.hpp#L16 is the original handmodel colors. openpose is under a non-free license - my personal opinion is using just the colors is fair use, but IANAL.

**Reply by hafriedlander** [reply]
Hmm. Those don't seem to line up with the colors in the reference image though 😕

**Comment by monsieurpooh** [comment]
What are the actual hex colors? The ones in the image are 7 digits long. But hex colors are either 6 digits long if no Alpha, 8 digits long with Alpha. What is the 7th digit doing? I thought about ignoring the 7th digit. But clearly the magenta of "gravel" is wrong in that case. It looks more like E040C0. Then what does E0400C0 mean? What color is it really?

**Reply by AugmentedRealityCat** [reply]
Thanks for flagging this mistake. I've now fixed it.
Here is a new version (v22) with proper HEX color codes. 
[T2i_Segmentation_Color_Ref_v22.pdf](https://github.com/user-attachments/files/21433657/T2i_Segmentation_Color_Ref_v22.pdf)
[T2i_Segmentation_Color_Ref_v22.xls](https://github.com/user-attachments/files/21433658/T2i_Segmentation_Color_Ref_v22.xls)
[T2i_Segmentation_Color_Ref_v22.ods](https://github.com/user-attachments/files/21433659/T2i_Segmentation_Color_Ref_v22.ods)

And here is a picture of it:
<img alt="T2i_Segmentation_Color_Ref_v22_full" src="https://github.com/user-attachments/assets/6d7d0984-0e82-477c-b489-565cfe2e9846" />



The wrong HEX values were induced by an error in the RGB to HEX code. It was using 3 digits for the B channel instead of 2. You can cross-check it if you load the XLS or ODS version. For the first item, the new code should be `=DEC2HEX(D6,2)&DEC2HEX(E6,2)&DEC2HEX(F6,2)` instead of the old (bad) `=DEC2HEX(D6,2)&DEC2HEX(E6,2)&DEC2HEX(F6,3)`.
